### PR TITLE
Remove runtime tests, simplify Rest vs CLI test hierarchy, add sniff test for all runtimes

### LINKED ISSUE #3845: WskRestOperation does not properly set the exec kind even when one is specified
The way the `create` is implemented, it will ignore an explicitly defined `kind` when it cannot infer it in the presence of a file artifact:
https://github.com/apache/incubator-openwhisk/blob/3ab604c167ac5bec76c756e3e76473863eb0a25d/tests/src/test/scala/common/rest/WskRestOperations.scala#L317

This is wrong - the provided kind should take precedence.

### PR DESCRIPTION
**Pull request by rabbah**
Additional cleanup of tests:

1. [x] removes WskRest* where possible and makes the base test suite the Rest tests
2. [x] removes runtime tests should should be moved to their respective runtimes
3. [x] runs the unicode tests over all the non-deprecated kinds per the runtime manifest

Also fixes #3845. See issue for details.

## Related issue and scope

- [x] I opened an issue to propose and discuss this change (#3737, #3845)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [x] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [x] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

**Comment by chetanmeh** [comment]
@rabbah Do the unicode test pass for you on Mac?. For me they always failed and only passed on travis. May be related to default encoding and we may need to make it more explicit

**Comment by rabbah** [comment]
They do yes - do you run the tests from gradle or ij?

**Comment by rabbah** [comment]
I just need a sniff tests - it doesn't have to be unicode (which should be covered by the unit proxy tests anyway). So it could be "echo" as well.

File: tests/src/test/scala/system/basic/WskBasicJavaTests.scala [hunk 1]
```diff
@@ -1,104 +0,0 @@
...
-  val wsk: WskOperations
-  val expectedDuration = 120.seconds
-  val activationPollDuration = 60.seconds
-
-  behavior of "Java Actions"
-
-  /**
-   * Test the Java "hello world" demo sequence
-   */
-  it should "Invoke a java action" in withAssetCleaner(wskprops) { (wp, assetHelper) =>
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicJavaTests.scala]: this test can be dropped.

File: tests/src/test/scala/system/basic/WskBasicJavaTests.scala [hunk 2]
```diff
@@ -1,104 +0,0 @@
...
-    withClue("Test duration exceeds expectation (ms)") {
-      val duration = System.currentTimeMillis() - start
-      duration should be <= expectedDuration.toMillis
-    }
-  }
-
-  /*
-   * Example from the docs.
-   */
-  it should "Invoke a Java action where main is in the default package" in withAssetCleaner(wskprops) {
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicJavaTests.scala]: move this to unit test.

File: tests/src/test/scala/system/basic/WskBasicJavaTests.scala [hunk 3]
```diff
@@ -1,104 +0,0 @@
...
-      assetHelper.withCleaner(wsk.action, name) { (action, _) =>
-        action.create(name, file, main = Some("Hello"))
-      }
-
-      withActivation(wsk.activation, wsk.action.invoke(name, Map()), totalWait = activationPollDuration) {
-        _.response.result.get.toString should include("Hello stranger!")
-      }
-  }
-
-  it should "Ensure that Java actions cannot be created without a specified main method" in withAssetCleaner(wskprops) {
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicJavaTests.scala]: move this to unit test.

File: tests/src/test/scala/system/basic/WskBasicNode6Tests.scala [hunk 4]
```diff
@@ -1,94 +0,0 @@
...
-abstract class WskBasicNode6Tests extends TestHelpers with WskTestHelpers with JsHelpers {
-
-  implicit val wskprops = WskProps()
-  val wsk: WskOperations
-  val defaultAction = Some(TestUtils.getTestActionFilename("hello.js"))
-  lazy val currentNodeJsKind = "nodejs:6"
-
-  behavior of s"Runtime $currentNodeJsKind"
-
-  it should "Ensure that NodeJS actions can have a non-default entrypoint" in withAssetCleaner(wskprops) {
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicNode6Tests.scala]: drop, duplicate of unit test.

File: tests/src/test/scala/system/basic/WskBasicNode6Tests.scala [hunk 5]
```diff
@@ -1,94 +0,0 @@
...
-      }
-
-      withActivation(wsk.activation, wsk.action.invoke(name)) { activation =>
-        val response = activation.response
-        response.result.get.fields.get("error") shouldBe empty
-        response.result.get.fields.get("greetings") should be(Some(JsString("Hello from a non-standard entrypoint.")))
-      }
-  }
-
-  it should "Ensure that zipped actions are encoded and uploaded as NodeJS actions" in withAssetCleaner(wskprops) {
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicNode6Tests.scala]: drop, duplicate of unit test.

File: tests/src/test/scala/system/basic/WskBasicNode6Tests.scala [hunk 6]
```diff
@@ -1,94 +0,0 @@
...
-      }
-
-      withActivation(wsk.activation, wsk.action.invoke(name)) { activation =>
-        val response = activation.response
-        response.result.get.fields.get("error") shouldBe empty
-        response.result.get.fields.get("author") shouldBe defined
-      }
-  }
-
-  it should "Ensure that returning an empty rejected Promise results in an errored activation" in withAssetCleaner(
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicNode6Tests.scala]: move this to unit test.

File: tests/src/test/scala/system/basic/WskBasicNodeDefaultTests.scala [hunk 7]
```diff
@@ -1,125 +0,0 @@
...
-abstract class WskBasicNodeDefaultTests extends TestHelpers with WskTestHelpers with JsHelpers {
-
-  implicit val wskprops = WskProps()
-  val wsk: WskOperations
-  val defaultAction = Some(TestUtils.getTestActionFilename("hello.js"))
-  val currentNodeJsDefaultKind = "nodejs:6"
-
-  behavior of "NodeJS default runtime"
-
-  it should "Map a kind of nodejs:default to the current default NodeJS runtime" in withAssetCleaner(wskprops) {
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicNodeDefaultTests.scala]: drop - covered by unit tests for exec manifest.

File: tests/src/test/scala/system/basic/WskBasicNodeDefaultTests.scala [hunk 8]
```diff
@@ -1,125 +0,0 @@
...
-      }
-
-      val result = wsk.action.get(name)
-      withPrintOnFailure(result) { () =>
-        val action = convertRunResultToJsObject(result)
-        action.getFieldPath("exec", "kind") should be(Some(currentNodeJsDefaultKind.toJson))
-      }
-  }
-
-  it should "Ensure that zipped actions cannot be created without a kind specified" in withAssetCleaner(wskprops) {
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicNodeDefaultTests.scala]: this should be a unit test against the cli.

File: tests/src/test/scala/system/basic/WskBasicNodeDefaultTests.scala [hunk 9]
```diff
@@ -1,125 +0,0 @@
...
-      val createResult = assetHelper.withCleaner(wsk.action, name, confirmDelete = false) { (action, _) =>
-        action.create(name, file, expectedExitCode = ANY_ERROR_EXIT)
-      }
-
-      val output = s"${createResult.stdout}\n${createResult.stderr}"
-
-      output should include("kind")
-  }
-
-  it should "Ensure that JS actions created with no explicit kind use the current default NodeJS runtime" in withAssetCleaner(
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicNodeDefaultTests.scala]: drop - or move to a cli unit test.

File: tests/src/test/scala/system/basic/WskBasicNodeDefaultTests.scala [hunk 10]
```diff
@@ -1,125 +0,0 @@
...
-    }
-  }
-
-  def convertRunResultToJsObject(result: RunResult): JsObject = {
-    val stdout = result.stdout
-    val firstNewline = stdout.indexOf("\n")
-    stdout.substring(firstNewline + 1).parseJson.asJsObject
-  }
-
-  it should "Ensure that NodeJS actions can have a non-default entrypoint" in withAssetCleaner(wskprops) {
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicNodeDefaultTests.scala]: drop, duplicate of unit test (also was repeated above).

File: tests/src/test/scala/system/basic/WskBasicNodeDefaultTests.scala [hunk 11]
```diff
@@ -1,125 +0,0 @@
...
-      }
-
-      withActivation(wsk.activation, wsk.action.invoke(name)) { activation =>
-        val response = activation.response
-        response.result.get.fields.get("error") shouldBe empty
-        response.result.get.fields.get("greetings") should be(Some(JsString("Hello from a non-standard entrypoint.")))
-      }
-  }
-
-  it should "Ensure that zipped actions are encoded and uploaded as NodeJS actions" in withAssetCleaner(wskprops) {
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicNodeDefaultTests.scala]: drop, duplicate.

File: tests/src/test/scala/system/basic/WskBasicPythonTests.scala [hunk 12]
```diff
@@ -1,142 +0,0 @@
...
-
-@RunWith(classOf[JUnitRunner])
-abstract class WskBasicPythonTests extends TestHelpers with WskTestHelpers with Matchers with JsHelpers {
-
-  implicit val wskprops = WskProps()
-  val wsk: WskOperations
-
-  behavior of "Native Python Action"
-
-  it should "invoke an action and get the result" in withAssetCleaner(wskprops) { (wp, assetHelper) =>
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicPythonTests.scala]: drop, duplicate.

File: tests/src/test/scala/system/basic/WskBasicPythonTests.scala [hunk 13]
```diff
@@ -1,142 +0,0 @@
...
-          main = Some("niam"),
-          kind = Some("python:default"))
-      }
-
-      withActivation(wsk.activation, wsk.action.invoke(name, Map("name" -> "Prince".toJson))) {
-        _.response.result.get shouldBe JsObject("greeting" -> JsString("Hello Prince!"))
-      }
-  }
-
-  Seq("python:2", "python:3").foreach { kind =>
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicPythonTests.scala]: drop, duplicate.

File: tests/src/test/scala/system/basic/WskBasicPythonTests.scala [hunk 14]
```diff
@@ -1,142 +0,0 @@
...
-    assetHelper.withCleaner(wsk.action, name) { (action, _) =>
-      action.create(name, Some(TestUtils.getTestActionFilename("hello.py")))
-    }
-
-    withActivation(wsk.activation, wsk.action.invoke(name, Map("name" -> "Prince".toJson))) {
-      _.response.result.get.toString should include("Prince")
-    }
-  }
-
-  it should "invoke an action with a non-default entry point" in withAssetCleaner(wskprops) { (wp, assetHelper) =>
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicPythonTests.scala]: drop, duplicate.

File: tests/src/test/scala/system/basic/WskBasicPythonTests.scala [hunk 15]
```diff
@@ -1,142 +0,0 @@
...
-    assetHelper.withCleaner(wsk.action, name) { (action, _) =>
-      action.create(name, Some(TestUtils.getTestActionFilename("niam.py")), main = Some("niam"))
-    }
-
-    withActivation(wsk.activation, wsk.action.invoke(name, Map())) {
-      _.response.result.get.fields.get("greetings") should be(Some(JsString("Hello from a non-standard entrypoint.")))
-    }
-  }
-
-  it should "invoke an action from a zip file with a non-default entry point" in withAssetCleaner(wskprops) {
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicPythonTests.scala]: drop, duplicate.

File: tests/src/test/scala/system/basic/WskBasicPythonTests.scala [hunk 16]
```diff
@@ -1,142 +0,0 @@
...
-        action.create(name, Some(TestUtils.getTestActionFilename("pythonVersion.py")), kind = Some(kind))
-      }
-
-      withActivation(wsk.activation, wsk.action.invoke(name)) {
-        _.response.result.get shouldBe JsObject("version" -> JsNumber(kind.takeRight(1).toInt))
-      }
-    }
-  }
-
-  it should "invoke an action and confirm expected environment is defined" in withAssetCleaner(wskprops) {
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicPythonTests.scala]: drop, duplicate.

File: tests/src/test/scala/system/basic/WskBasicPythonTests.scala [hunk 17]
```diff
@@ -1,142 +0,0 @@
...
-      withActivation(wsk.activation, wsk.action.invoke(name)) { activation =>
-        val result = activation.response.result.get
-        result.fields.get("error") shouldBe empty
-        result.fields.get("auth") shouldBe Some(
-          JsString(WhiskProperties.readAuthKey(WhiskProperties.getAuthFileForTesting)))
-        result.fields.get("edge").toString.trim should not be empty
-      }
-  }
-
-  it should "invoke an invalid action and get error back" in withAssetCleaner(wskprops) { (wp, assetHelper) =>
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicPythonTests.scala]: duplicate - but unit test doesn't check error message as done here, should tighten unit test accordingly.

File: tests/src/test/scala/system/basic/WskBasicPythonTests.scala [hunk 18]
```diff
@@ -1,142 +0,0 @@
...
-    withActivation(wsk.activation, wsk.action.invoke(name)) { activation =>
-      activation.response.result.get.fields.get("error") shouldBe Some(
-        JsString("The action failed to generate or locate a binary. See logs for details."))
-      activation.logs.get.mkString("\n") should { not include ("pythonaction.py") and not include ("flask") }
-    }
-  }
-
-  behavior of "Python virtualenv"
-
-  Seq(("python2_virtualenv.zip", "python:2"), ("python3_virtualenv.zip", "python:3")).foreach {
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicPythonTests.scala]: drop, duplicates.

File: tests/src/test/scala/system/basic/WskBasicSwift3Tests.scala [hunk 19]
```diff
@@ -1,65 +0,0 @@
...
-  implicit val wskprops: common.WskProps = WskProps()
-  val wsk: WskOperations
-  val defaultAction: Some[String] = Some(TestUtils.getTestActionFilename("hello.swift"))
-  lazy val actionKind = "swift:3.1.1"
-  val activationMaxDuration = 2.minutes
-  val activationPollDuration = 3.minutes
-
-  behavior of "Swift runtime"
-
-  it should "Ensure that Swift actions can have a non-default entrypoint" in withAssetCleaner(wskprops) {
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskBasicSwift3Tests.scala]: drop, duplicate.

File: tests/src/test/scala/system/basic/WskRestActionTests.scala [hunk 20]
```diff
@@ -1,38 +0,0 @@
...
-import common.rest.WskRestOperations
-import org.junit.runner.RunWith
-import org.scalatest.junit.JUnitRunner
-import spray.json._
-
-@RunWith(classOf[JUnitRunner])
-class WskRestActionTests extends WskActionTests with WskActorSystem {
-  override val wsk = new WskRestOperations
-
-  it should "create an action with an empty file" in withAssetCleaner(wskprops) { (wp, assetHelper) =>
```
Comment (rabbah) [inline tests/src/test/scala/system/basic/WskRestActionTests.scala]: The docker skeleton and python runtimes accept an empty code string during init, and fail in run (docker skeleton will just run the stub, and python will return error = action did not return dictionary). 

Java will fail with class/method not found for main during class loader.

Swift will fail with did not generate an executable.

Dropping this test is OK - I don't think we're getting any increased coverage. It would be nice to make the failures uniform but this may be a little difficult for all the runtimes to normalize.

See #3839 and #3844.

**Comment by dgrove-oss** [comment]
PG1 3103 🏃

**Comment by markusthoemmes** [comment]
Thanks a big bunch for all of this grunt work 👍
